zing_stats.py

```diff
@@ -164,140 +180,212 @@ def main():
-            log.info('Wrote %s for %s group', file_path, group)
-    else:
+    all_projects = teams_map['All']
+    for team in sorted(teams_map):
+        team_projects = teams_map[team]
+        teams = teams_map.keys()
+        html = generate_html(args, df, num_changes, start_dt, finish_dt,
+                             team_projects, all_projects, team, teams)
         dir_path = os.path.join(args.output_dir, file_prefix)
-        file_name = 'index.html'
+        if team == 'All':
+            file_name = 'index.html'
+        else:
+            file_name = '%s.html' % re.sub(r'\W+', '_', team.lower())
         file_path = os.path.join(dir_path, file_name)
         if not os.path.exists(dir_path):
             os.makedirs(dir_path)
-        html = generate_html(args, df, num_changes, start_dt, finish_dt,
-                             df.keys(), all_projects, 'All')
         with open(file_path, 'w') as f:
             f.write(html)
-        log.info('Wrote %s', file_path)
+        log.info('Wrote %s for team: "%s"', file_path, team)
+
+
+def gather_github_prs(args, oldest_timestamp, projects):
+    projects_github = [x['name'] for x in projects.get('github', dict())]
+
+    if len(projects_github) < 1:
+        return 0, dict()
+
+    log.info('Gathering PRs from %s for %s', args.github_url,
+             ', '.join(sorted(projects_github)))
+    total_prs = 0
+    prs = dict()
+    for project in projects_github:
+        prs[project] = dict()
+
+        project_finished = False
+        next_page = True
+
+        payload = {'state': 'all', 'sort': 'updated', 'direction': 'dsc'}
+        query = ('%s/api/v3/repos/%s/pulls'
+                 % (args.github_url, project))
+        while next_page:
+            if args.github_user and args.github_token:
+                response = requests.get(
+                    query, params=payload,
+                    verify=args.verify_https_requests,
+                    auth=(args.github_user, args.github_token))
+            else:
+                response = requests.get(query, params=payload,
+                                        verify=args.verify_https_requests)
+            log.debug(response.url)
+            if response.status_code == 404:
+                if args.github_user and args.github_token:
+                    log.error('Skipping %s (404 while listing PRs, the '
+                              '--github-user and --github-token specified '
+                              'don\'t have access to this project)', project)
+                else:
+                    log.error('Skipping %s (404 while listing PRs, try '
+                              'providing a --github-user and --github-token '
+                              'with access to this project)', project)
+                prs.pop(project)
+                break
+            results = response.json()
+            log.debug(json.dumps(results, sort_keys=True, indent=4,
+                                 separators=(',', ': ')))
```

Comment:
Is this a case for some functools and helper function to have log.debug() only resolve the json.dumps() if needed? Otherwise I'm guessing python current expands the method results before passing a reference to log.debug and then discarding if debug logging is not enabled.

Reply:
Good optimisation, adding a TODO for this